### ISSUE TITLE
only default cert matches signature_algorithm extension

### ISSUE THREAD

**Issue by mlibbey** [apache/trafficserver#2482]
When ATS uses its non default cert, and the client uses the signature_algorithms TLS extention, ATS returns a mismatch of the signature algorithm used in the ServerKeyExchange. 

 [RFC 5246 section (7.4.3)](https://tools.ietf.org/html/rfc5246#section-7.4.3) says:
    "If the client has offered the "signature_algorithms" extension, the signature algorithm and hash algorithm MUST be a pair listed in that extension [the ServerKeyExchange].”

This is violated by the server. Thus, the client must return an (Invalid Parameter) alert. There are client TLS stacks that have explicit checks for this requirement.

Here's how to reproduce:

- when ssl_multicert is
```
dest_ip=* ssl_cert_name=mycert.example.com
```

then the signature_algorithm extension matches:
```
$ d="mycert.example.com"; 
alg=RSA; 
for hash in SHA1 SHA256 SHA384 SHA512; do 
  result=$(echo | openssl s_client -connect myserver.example.com:443 \
-servername $d -sigalgs ${alg}+${hash} -no_ticket 2>/dev/null | grep digest); 
  echo -n $result; 
  if [[ $result == *"$hash"* ]]; 
  then echo " (success)";
  else echo " (fail)"; \
  fi; 
done
Peer signing digest: SHA1 (success)
Peer signing digest: SHA256 (success)
Peer signing digest: SHA384 (success)
Peer signing digest: SHA512 (success)
```

- if the ssl _multicert is 
```
 ssl_cert_name=mycert.example.com 
```
(e.g., no "dest_ip=*")

then the signature_algorithm extension doesn't match (this machine no longer has this config):
```
$ d="mycert.example.com"; 
alg=RSA; 
for hash in SHA1 SHA256 SHA384 SHA512; do 
  result=$(echo | openssl s_client -connect myserver.example.com:443 \
 -servername $d -sigalgs ${alg}+${hash} -no_ticket 2>/dev/null | grep digest); 
  echo -n $result;
 if [[ $result == *"$hash"* ]]; 
 then echo " (success)"; 
 else echo " (fail)"; 
 fi; 
done
Peer signing digest: SHA1 (success)
Peer signing digest: SHA1 (fail)
Peer signing digest: SHA1 (fail)
Peer signing digest: SHA1 (fail)
```
the presence of any other cert does not  affect this behavior -- e.g., the lines above can be the only line in ssl_multicert.

Tested in 7.1.1; 7.1, and 6.1.2.

**Comment by masaori335** [comment]
Masakazu and I investigated this issue and found root cause. Let me describe what is happening.

1. Accept SSL Client Hello 
When `SSL_accept()` is called, the signature algorithms extension offered by client is proceeded in `tls1_set_server_sigalgs()` and the hash algorithm is stored as `digest` in the `cert` of SSL object ( `ssl->cert->pkeys[idx].digest` )
https://github.com/openssl/openssl/blob/OpenSSL_1_0_2l/ssl/t1_lib.c#L3903

2. SSL Cert Callback
ATS set `ssl_cert_callback()` as the `cer_cb()`, so `ssl_cert_callback()` is called.
SSL_CTX is lookuped by SNI and it is set to SSL object by `SSL_set_SSL_CTX()`.

3. Set SSL_CTX
If `dest_ip=*` is in ssl_multicert.config, the SSL_CTX is same to SSL_CTX which SSL object already have, so `SSL_set_SSL_CTX()` do nothing.
But when the lookuped SSL_CTX is different from default SSL_CTX, the `cert` of SSL object is ovewritten by `cert` of SSL_CTX. Thus stored hash algorithms offered by client in the `cert` is also overwritten by default value (SHA1).

FWIW, we also noticed that this issue is not happened with OpenSSL-1.1.*, because hash algorithms are stored in `s3` of SSL object.
https://github.com/openssl/openssl/blob/OpenSSL_1_1_0/ssl/t1_lib.c#L3442

**Comment by maskit** [comment]
This might be a regression?
https://github.com/openssl/openssl/commit/4e05aedbcab7f7f83a887e952ebdcc5d4f2291e4

**Comment by maskit** [comment]
Similar fix on 1.0.2 branch.
https://github.com/openssl/openssl/commit/1ce95f19601bbc6bfd24092c76c8f8105124e857

**Comment by masaori335** [comment]
For 1.0.2, `shared_sigalgs` also should be copied?

**Comment by maskit** [comment]
This is probably the original report.
https://bugzilla.redhat.com/show_bug.cgi?id=1150033

**Comment by masaori335** [comment]
It looks like v1.0.2 should copy CERT_PKEYs like 1.0.1 *or* setup CERT_PKEYs from peer_sigalgs again.

**Comment by maskit** [comment]
I agree. Probably copying CERT_PKEYs is better option to have consistency. I guess the fix for 1.0.2 was just cherry-picked and nobody tested actually.

**Comment by masaori335** [comment]
This is fixed in upstream and the fix is released by OpenSSL 1.0.2m.
